### ISSUE TITLE
DataLoaderDispatcherInstrumentation & unregister dataloader

### ISSUE THREAD

**Issue by joesankey** [graphql-java/graphql-java#757]
I realise the data-loader integration is new and maybe DataLoaderDispatcherInstrumentation is still a work in progress.

Wondering what the plan is for how a dataloader gets unregistered at the end of a request.

For example say the application needs to create new dataloader per request, to prevent sharing caches across requests. What is the best approach to unregister a dataloader, from the DataLoaderDispatcherInstrumentation, when the request is complete?

**Comment by szantogab** [comment]
Good question, furthermore, I still don't get it how the data loader can be used with JPA. Like when a user has many posts, and the client wants to retrieve all the users, and their posts, how can it be batched efficiently in two queries? (We obviously need to query the users, and then somehow get the IDs of each user's posts, and then load it with dataloader? Makes no sense.)

Anyone to explain? :)

**Comment by joesankey** [comment]
@szantogab you need to use BatchedExecutionStrategy, which will parse the graph and resolve Types in batches for you.

**Comment by szantogab** [comment]
@joesankey thanks, do you have any examples? I've yet to find any.

**Comment by bbakerman** [comment]
@joesankey - you have a couple of options.  You can hold onto the DataLoaderRegistry and unload keys but better yet throw it all away per request.

You can create a per request DataLoaderDispatcherInstrumentation and a per request Graphql object.

Creating a new `GraphQL` object per request is super  cheap.  The new master code has a trandform object for making this even easier.

     GraphQL perrequestGraphQL = previousGraphql.transform( builder  -> builder.setInstrumentation(...));

Creating the schema (especially if you use IDL) is more expensive and essentially static.  So I would keep the schema as a static variable (or injected IoC component) and recreate your graphql object each request including data loader support

**Comment by bbakerman** [comment]
@szantogab - your dont have to use BatchedExecutionStrategy - in fact I actively recommend against it since its not 100% graphql spec compliant (its ok but not 100%)

See https://medium.com/@bradbaker/big-movements-in-the-graphql-java-world-67629fd45508 for an example of using data fetcher to gather N requests into one.

The key behind data fetcher is that if you ask for key K1 and later K2 and K3 and so on, it holds requests until a suitable time and presents K1,K2,K3 to a batch loading function.

They way you load V1,V2,V3 in one go and not 3.  It also caches the previous results (options to turn that off) and hence if you ever ask for K1 again it immediately can give you a past result.

This has some example code here -

https://github.com/graphql-java/graphql-java/blob/master/src/test/groovy/example/http/HttpMain.java#L166

**Comment by joesankey** [comment]
@bbakerman interesting, now that we have the Asynchronous strategy, what is the advantage of BatchedExecutionStrategy?

**Comment by bbakerman** [comment]
With Asyncrhonous strategies and DataLoader there really isnt any need for the older BatchedExecutionStrategy.

Its is not 100% graphql compliant (for example null values for non null fields does not work) and its a very early attempt at batching.

Facebook proved batching can work via [dataloader](https://github.com/facebook/dataloader) and java-dataloader is a faithful port of that proven pattern.

In short there is no advantage to BatchedExecutionStrategy and I recommend you dont use it

**Comment by szantogab** [comment]
@bbakerman Thanks for your reply. I think I got the whole dataloader concept, but I still have a problem with it:

Let's assume we have a set of users, and each user can have multiple posts. How can we efficiently query all the users and their posts while using a relational database?

Here's how the query would look like:
query {
    users {
        id
        posts {
            id
            title
        }
    }
}

Please correct me where my solution is incorrect:

1. Query all the users from the db
2. With dataloader, we `prime` all the users with their IDs in the `userDataLoader`.
3. In the user.posts data fetcher, we use a `userPostLoader` and `load` a given user's id, to get back its posts.

This should result in a db query that fetches all the users, and another query that fetches all of the users posts at once.

But it doesn't, because unfortunately `DataLoaderDispatcherInstrumentation` calls `dispatchAll()` too early, when there is only one userId has been `load`ed to `userPostLoader`

Where am I going wrong here?
